[Server] Use same Golang logger and log format

haanhvu opened the pull request:
Signed-off-by: haanhvu <haanh6594@gmail.com>

**Description**

This PR fixes #5579 

**Notes for Reviewers**


**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 
From what I read in https://github.com/meshery/meshery/issues/5579#issuecomment-1149521411 we need to replace the use of `github.com/sirupsen/logrus` with the use of `github.com/layer5io/meshkit/logger`.

I did that and the logs have the same format now:
![Screenshot from 2022-06-12 20-09-06](https://user-images.githubusercontent.com/75315486/173234779-174cfb52-78da-4ac1-bcc7-62e85f62f86d.png)

However, there're still some uses of `github.com/sirupsen/logrus` left because I can't find alternatives from `github.com/layer5io/meshkit/logger`. If we want to totally get rid of `github.com/sirupsen/logrus`, I think we need to expand `github.com/layer5io/meshkit/logger`.

WDYT?

leecalcote commented:
Thank you, @haanhvu. The reformatted logs (swapped loggers) looks great.

Do you have examples of what the `meshkit/logger` will need to be expanded to support? Your suggestion sounds about right to me.

haanhvu commented:
> Do you have examples of what the `meshkit/logger` will need to be expanded to support? Your suggestion sounds about right to me.

@leecalcote the methods from `sirupsen/logrus` that I can't find alternatives in `meshkit/logger` are `Fatal()`, `Fatalf()`, and `Error()` of multiple parameters (like [this](https://github.com/meshery/meshery/blob/358b1360c788b60da719846e2905ad4be0487754/cmd/main.go#L198)). Also, `logrus.SetLevel()` and `logrus.GetLevel()` don't have alternatives in `meshkit/logger` but I don't think these two methods affect the log format?

So either we need to expand the definitions of `Fatal()`, `Fatalf()`, and `Error()` of multiple parameters in `meshkit/logger`, or they already exist somewhere but I didn't see^^

Aisuko reviewed (dismissed):
I believe it is ok we migrate to meshkit/logger.  Be careful the issues in static checking and CI process.

MUzairS15 commented:
https://github.com/meshery/meshery/blob/358b1360c788b60da719846e2905ad4be0487754/cmd/main.go#L198

Can't we do
```
er := errors.New(providerurl + "is invalid url skipping provider")
log.Error(er)

MUzairS15 reviewed (commented):
Please fix lint issues.

leecalcote commented:
@haanhvu heads-up: a merge conflict.

leecalcote commented:
Conflict resolved. Syncing with `master`

gr455 commented on cmd/main.go:
```diff
@@ -195,7 +196,8 @@ func main() {
 		parsedURL, err := url.Parse(providerurl)
 		if err != nil {
-			logrus.Error(providerurl, "is invalid url skipping provider")
+			er := errors.New(providerurl + "is invalid url skipping provider")
```
Might want to make this a meshkit error from `github.com/layer5io/meshkit/errors`
because https://docs.meshery.io/project/contributing/contributing-error#warning

leecalcote replied:
Nice. I don’t know that @haanhvu has been introduced to MeshKit Errors, so this is a good note.

MUzairS15 replied:
yes, @gr455 errors.New(..) short des, long des, cause, remedy... .
@haanhvu Should create a new error in errors.go file and use this error in logger.

gr455 replied:
Yep, just like Uzair said. You can refer other places in the codebase like [this one in mesheryctl](https://github.com/meshery/meshery/tree/master/mesheryctl/internal/cli/root) to get the structure of the errors.go file. All the best!

Revolyssup replied:
@haanhvu  Remember that throwing non-meshkit errors in meshkit logger will cause panic. Thank you @gr455  for noticing, I totally missed this.

haanhvu replied:
yes let me see. Thanks you guys for the info.

haanhvu replied:
Hey guys, about the new error, I'll create it with `meshkit/errors`.

However, there are existing errors that are not meshkit errors. Should I keep them handled by `logrus`? For example, [this](https://github.com/meshery/meshery/pull/5603/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6R87-R89).

haanhvu commented:
Sorry for a little delay. I've just been through a roller coaster week^^ Will come back to this PR tomorrow.

haanhvu commented on cmd/error.go:
```diff
@@ -0,0 +1,56 @@
+import "github.com/layer5io/meshkit/errors"
+
+const (
+	ErrRegisteringMesheryOAMTraitsCode = "code1"
+	ErrRegisteringMesheryOAMWorkloadsCode = "code2"
+	ErrRetrievingUserHomeDirectoryCode = "code3"
+	ErrCreatingUserDataDirectoryCode = "code4"
+	ErrCreatingMapPreferencePersisterInstanceCode = "code5"
+	ErrDatabaseAutoMigrationCode = "code6"
+	ErrInvalidURLSkippingProviderCode = "code7"
+	ErrListenAndServeCode = "code8"
+	ErrCleaningUpLocalProviderCode = "code9"
+	ErrClosingDatabaseInstanceCode = "code10"
```
@gr455 yeah I'll try that

haanhvu replied:
@gr455 yeah I read in the doc that the codes can be whatever and the CI will take care of that? Let me try the command you suggested.

haanhvu commented:
I migrated all errors to meshkit errors for meshkit logger. Also editted some commands  in Makefile and Github Action to make the new error.go defined.

Is there anything else needs to be done? There're some check failures but they don't seem to be in this PR's scope.

haanhvu commented:
@Revolyssup pls take a look when you have time

leecalcote commented:
@gr455 thank you for reviewing with @haanhvu

gr455 reviewed (commented):
LGTM!

haanhvu commented:
Almost forgot to mention, should these lines be deleted? The nearest error's already handled above.
https://github.com/meshery/meshery/blob/0dbac4aa74e17ef342d29facbbe5e090d7791aa0/cmd/main.go#L147_L149

gr455 commented:
> 

Yep seems redundant

leecalcote commented:
Good eye, @haanhvu

haanhvu commented:
> any updates here @haanhvu ?

I deleted reduntdant lines as agreed with Ruturaj [here](https://github.com/meshery/meshery/pull/5603#issuecomment-1173077173)

For `logrus.Fatal()`, I replaced it with calling meshkit `log.error()` on fatal meshkit errors then `os.Exit(1)`. I asked if that's ok [here](https://github.com/meshery/meshery/pull/5603/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6R99-R100)

Seeing nothing else to do. The logger runs fine:
![Screenshot from 2022-07-17 20-50-04](https://user-images.githubusercontent.com/75315486/179401517-b787feac-049d-4b88-9a0f-55778c23361f.png)

gr455 reviewed (dismissed):
@leecalcote Smells like it's done :+1: 
Failing tests are not related and Aadhitya is working on fixing them

gr455 commented:
Oh by the way @haanhvu. What about [these](https://github.com/meshery/meshery/pull/5603/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6L50). Can we replace these with meshkit too?

haanhvu commented:
> Oh by the way @haanhvu. What about [these](https://github.com/meshery/meshery/pull/5603/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6L50). Can we replace these with meshkit too?

@gr455 the link doesn't point to any line but I guess you're mentioning [this](https://github.com/meshery/meshery/blob/55742d8f9540ff9f4f304f9018370adc16df851d/cmd/main.go#L50)

Yeah sure we can replace this too. I didn't replace this because currently it's a fatal error happening before the initialization of meshkit logger. It's actually the first error too. So I guess the code writer means if the server has to die it should die as soon as possible (no need to wait for the initialization of meshkit logger)?

If we want to replace it with meshkit error, we can move the initialization of meshkit logger before it.

gr455 commented:
> > Oh by the way @haanhvu. What about [these](https://github.com/meshery/meshery/pull/5603/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6L50). Can we replace these with meshkit too?
> 
> @gr455 the link doesn't point to any line but I guess you're mentioning [this](https://github.com/meshery/meshery/blob/55742d8f9540ff9f4f304f9018370adc16df851d/cmd/main.go#L50)

Yep @haanhvu that's the one I meant

> Yeah sure we can replace this too. I didn't replace this because currently it's a fatal error happening before the initialization of meshkit logger. It's actually the first error too. So I guess the code writer meant if the server has to die it should die as soon as possible (no need to wait for the initialization of meshkit logger)?
> 
> If we want to replace it with meshkit error, we can move the initialization of meshkit logger to happen before it.

Yeah, I think that would be a good idea, the error doesn't seem like something that should be immediately caught and killed. I think it can wait for the logger to be initialized. So you can move the logger initialization above the UUID function call.

That's all. We can get it merge right after that, what do you say?

haanhvu commented:
@gr455 yeah let me quickly replace it then we can get this done

haanhvu commented:
@gr455 done

gr455 reviewed (dismissed):
Done for real now! @leecalcote Let's merge this

haanhvu commented:
@Revolyssup @leecalcote @gr455 I was wondering why you didn't answer. But turned out my reviews were pending...
![Screenshot 2022-08-03 160931](https://user-images.githubusercontent.com/75315486/182570925-811bca68-7cdf-41ed-bea4-495834ec7b24.png)
![Screenshot 2022-08-03 161029](https://user-images.githubusercontent.com/75315486/182570974-c648352a-a155-4505-886a-1d8d16ffd42d.png)

So please continue the review from here.

Also, does anyone know how to fix this? After a quick Google search, I found this may be a common problem: https://github.com/microsoft/vscode-pull-request-github/issues/2822

gr455 commented:
> @Revolyssup @leecalcote @gr455 I was wondering why you didn't answer. But turned out my reviews were pending forever...
> 
> ![Screenshot 2022-08-03 160931](https://user-images.githubusercontent.com/75315486/182570925-811bca68-7cdf-41ed-bea4-495834ec7b24.png)
> 
> ![Screenshot 2022-08-03 161029](https://user-images.githubusercontent.com/75315486/182570974-c648352a-a155-4505-886a-1d8d16ffd42d.png)
> 
> So please continue the reviews from here.
> 
> Also, does anyone know how to fix this? All my review comments were pending until the reviewed code is outdated.
> 
> After a quick Google search, I found this may be a common problem and may happen to other contributors too: [microsoft/vscode-pull-request-github#2822](https://github.com/microsoft/vscode-pull-request-github/issues/2822)

`log.Error` seems fine to me :+1:

Revolyssup commented:
Thank you @haanhvu for this.
